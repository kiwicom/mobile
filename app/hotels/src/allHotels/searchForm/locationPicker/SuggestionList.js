// @flow

import * as React from 'react';
import { FlatList, View, Keyboard } from 'react-native';
import idx from 'idx';
import { StyleSheet } from '@kiwicom/mobile-shared';
import { createFragmentContainer, graphql } from '@kiwicom/mobile-relay';

import SuggestionListItem from './SuggestionListItem';
import type { SuggestionList_data } from './__generated__/SuggestionList_data.graphql';

type Props = {|
  data: SuggestionList_data,
  onCitySelected: (cityId: string | null, cityName: ?string) => void,
  search: string,
|};

type ListItemType = {|
  node: {
    id: string,
    name: ?string,
  },
|};

const styles = StyleSheet.create({
  listHeader: {
    marginTop: 20,
  },
});

export class SuggestionList extends React.Component<Props> {
  keyExtractor = (item: ListItemType) => {
    const city = idx(item, _ => _.node) || {};
    return city.id;
  };

  renderItem = ({ item }: { item: ListItemType }) => {
    const city = idx(item, _ => _.node);

    if (city == null) {
      return null;
    }

    return (
      <SuggestionListItem
        city={city}
        onCitySelected={this.props.onCitySelected}
        search={this.props.search}
      />
    );
  };

  render = () => (
    <FlatList
      data={idx(this.props.data, _ => _.hotelCities.edges)}
      renderItem={this.renderItem}
      keyExtractor={this.keyExtractor}
      ListHeaderComponent={<View style={styles.listHeader} />}
      onScroll={Keyboard.dismiss}
      keyboardShouldPersistTaps="handled"
    />
  );
}

export default createFragmentContainer(SuggestionList, {
  data: graphql`
    fragment SuggestionList_data on RootQuery
      @argumentDefinitions(prefix: { type: "String" }) {
      hotelCities(prefix: $prefix, first: 50)
        @connection(key: "SuggestionList_hotelCities") {
        edges {
          node {
            id
            name
          }
        }
      }
    }
  `,
});
