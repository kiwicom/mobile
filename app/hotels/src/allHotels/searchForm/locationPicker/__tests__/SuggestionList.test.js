// @flow

import * as React from 'react';
import { PlaygroundRenderer } from '@kiwicom/react-native-app-playground';

import { SuggestionList } from '../SuggestionList';

// $FlowRelayIssue: https://github.com/facebook/relay/issues/2394
const data = {
  hotelCities: {
    edges: [],
  },
};

for (let i = 1; i < 51; i++) {
  data.hotelCities.edges.push({
    node: {
      id: `${i}`,
      name: `City mock ${i}`,
    },
  });
}

function noop() {}

describe('SuggestionList', () => {
  it('renders', () => {
    PlaygroundRenderer.render(
      <SuggestionList data={data} search="" onCitySelected={noop} />,
    );
  });
});
