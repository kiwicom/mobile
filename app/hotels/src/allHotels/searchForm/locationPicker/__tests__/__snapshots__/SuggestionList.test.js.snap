// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestionList renders 1`] = `
<FlatList
  ListHeaderComponent={
    <Component
      style={
        Object {
          "marginTop": 5,
        }
      }
    />
  }
  data={
    Array [
      Object {
        "node": Object {
          "id": "1",
          "name": "City mock 1",
        },
      },
      Object {
        "node": Object {
          "id": "2",
          "name": "City mock 2",
        },
      },
      Object {
        "node": Object {
          "id": "3",
          "name": "City mock 3",
        },
      },
      Object {
        "node": Object {
          "id": "4",
          "name": "City mock 4",
        },
      },
      Object {
        "node": Object {
          "id": "5",
          "name": "City mock 5",
        },
      },
      Object {
        "node": Object {
          "id": "6",
          "name": "City mock 6",
        },
      },
      Object {
        "node": Object {
          "id": "7",
          "name": "City mock 7",
        },
      },
      Object {
        "node": Object {
          "id": "8",
          "name": "City mock 8",
        },
      },
      Object {
        "node": Object {
          "id": "9",
          "name": "City mock 9",
        },
      },
      Object {
        "node": Object {
          "id": "10",
          "name": "City mock 10",
        },
      },
      Object {
        "node": Object {
          "id": "11",
          "name": "City mock 11",
        },
      },
      Object {
        "node": Object {
          "id": "12",
          "name": "City mock 12",
        },
      },
      Object {
        "node": Object {
          "id": "13",
          "name": "City mock 13",
        },
      },
      Object {
        "node": Object {
          "id": "14",
          "name": "City mock 14",
        },
      },
      Object {
        "node": Object {
          "id": "15",
          "name": "City mock 15",
        },
      },
      Object {
        "node": Object {
          "id": "16",
          "name": "City mock 16",
        },
      },
      Object {
        "node": Object {
          "id": "17",
          "name": "City mock 17",
        },
      },
      Object {
        "node": Object {
          "id": "18",
          "name": "City mock 18",
        },
      },
      Object {
        "node": Object {
          "id": "19",
          "name": "City mock 19",
        },
      },
      Object {
        "node": Object {
          "id": "20",
          "name": "City mock 20",
        },
      },
      Object {
        "node": Object {
          "id": "21",
          "name": "City mock 21",
        },
      },
      Object {
        "node": Object {
          "id": "22",
          "name": "City mock 22",
        },
      },
      Object {
        "node": Object {
          "id": "23",
          "name": "City mock 23",
        },
      },
      Object {
        "node": Object {
          "id": "24",
          "name": "City mock 24",
        },
      },
      Object {
        "node": Object {
          "id": "25",
          "name": "City mock 25",
        },
      },
      Object {
        "node": Object {
          "id": "26",
          "name": "City mock 26",
        },
      },
      Object {
        "node": Object {
          "id": "27",
          "name": "City mock 27",
        },
      },
      Object {
        "node": Object {
          "id": "28",
          "name": "City mock 28",
        },
      },
      Object {
        "node": Object {
          "id": "29",
          "name": "City mock 29",
        },
      },
      Object {
        "node": Object {
          "id": "30",
          "name": "City mock 30",
        },
      },
      Object {
        "node": Object {
          "id": "31",
          "name": "City mock 31",
        },
      },
      Object {
        "node": Object {
          "id": "32",
          "name": "City mock 32",
        },
      },
      Object {
        "node": Object {
          "id": "33",
          "name": "City mock 33",
        },
      },
      Object {
        "node": Object {
          "id": "34",
          "name": "City mock 34",
        },
      },
      Object {
        "node": Object {
          "id": "35",
          "name": "City mock 35",
        },
      },
      Object {
        "node": Object {
          "id": "36",
          "name": "City mock 36",
        },
      },
      Object {
        "node": Object {
          "id": "37",
          "name": "City mock 37",
        },
      },
      Object {
        "node": Object {
          "id": "38",
          "name": "City mock 38",
        },
      },
      Object {
        "node": Object {
          "id": "39",
          "name": "City mock 39",
        },
      },
      Object {
        "node": Object {
          "id": "40",
          "name": "City mock 40",
        },
      },
      Object {
        "node": Object {
          "id": "41",
          "name": "City mock 41",
        },
      },
      Object {
        "node": Object {
          "id": "42",
          "name": "City mock 42",
        },
      },
      Object {
        "node": Object {
          "id": "43",
          "name": "City mock 43",
        },
      },
      Object {
        "node": Object {
          "id": "44",
          "name": "City mock 44",
        },
      },
      Object {
        "node": Object {
          "id": "45",
          "name": "City mock 45",
        },
      },
      Object {
        "node": Object {
          "id": "46",
          "name": "City mock 46",
        },
      },
      Object {
        "node": Object {
          "id": "47",
          "name": "City mock 47",
        },
      },
      Object {
        "node": Object {
          "id": "48",
          "name": "City mock 48",
        },
      },
      Object {
        "node": Object {
          "id": "49",
          "name": "City mock 49",
        },
      },
      Object {
        "node": Object {
          "id": "50",
          "name": "City mock 50",
        },
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  keyboardShouldPersistTaps="handled"
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReachedThreshold={2}
  onScroll={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;
